crypto/merklekeystore/keystore.go

```diff
@@ -0,0 +1,264 @@
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package merklekeystore
+
+import (
+	"errors"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+type (
+	// CommittablePublicKey is a key tied to a specific round and is committed by the merklekeystore.Signer.
+	CommittablePublicKey struct {
+		_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+		VerifyingKey crypto.VerifyingKey `codec:"pk"`
+		Round        uint64              `codec:"rnd"`
+	}
+
+	//Proof represent the merkle proof in each signature.
+	Proof merklearray.Proof
+
+	// Signature is a byte signature on a crypto.Hashable object,
+	// crypto.VerifyingKey and includes a merkle proof for the key.
+	Signature struct {
+		_struct              struct{} `codec:",omitempty,omitemptyarray"`
+		crypto.ByteSignature `codec:"bsig"`
+
+		Proof        Proof               `codec:"prf"`
+		VerifyingKey crypto.VerifyingKey `codec:"vkey"`
+	}
+
+	// Signer is a merkleKeyStore, contain multiple keys which can be used per round.
+	// Signer will generate all keys in the range [A,Z] that are divisible by some divisor d.
+	// in case A equals zero then signer will generate all keys from (0,Z], i.e will not generate key for round zero.
+	Signer struct {
+		_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+		// these keys should be temporarily stored in memory until Persist is called,
+		// in which they will be dumped into database and disposed of.
+		// non-exported fields to prevent msgpack marshalling
+		signatureAlgorithms []crypto.SignatureAlgorithm
+		keyStore            PersistentKeystore
+
+		// the first round is used to set up the intervals.
+		FirstValid uint64 `codec:"rnd"`
+
+		Interval uint64 `codec:"iv"`
+
+		Tree merklearray.Tree `codec:"tree"`
+	}
+
+	// Verifier is used to verify a merklekeystore.Signature produced by merklekeystore.Signer.
+	// it validate a merklekeystore.Signature by validate the commitment on the VerifyingKey and validate the signature with that key
+	Verifier struct {
+		_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+		Root [KeyStoreRootSize]byte `codec:"r"`
+
+		// indicates that this verifier corresponds to a specific array of ephemeral keys.
+		// this is used to distinguish between an empty structure, and nothing to commit to.
+		HasValidRoot bool `codec:"vr"`
```

Comment:
I believe `HasValidRoot` is meant to indicate whether the struct is the zero value or not, but it’s not used everywhere, since there are frequent instances of `== merklekeystore.Verifier{}`. If this is the case, I recommend creating an `IsZero` method on `merklekeystore.Verifier` that returns `HasValidRoot` and making this the only zero check in use.